File: python_dsa/trees/bst/impl.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class BST:
    def __init__(self):
        self.root = None

    def insert(self, value):
        new_node = TreeNode(value)

        if self.is_empty():
            self.root = new_node
            return

        current = self.root

        while current:
            if value < current.value:
                if current.left:
                    current = current.left
                else:
                    current.left = new_node
                    break
            elif value > current.value:
                if current.right:
                    current = current.right
                else:
                    current.right = new_node
                    break
            else:
                break

    def search(self, value):
        if self.is_empty():
            return None

        current = self.root

        while current:
            if value == current.value:
                return current
            else:
                if value < current.value:
                    current = current.left
                else:
                    current = current.right

    def in_order(self, current=None, result=None):
        if current is None:
            current = self.root
        if result is None:
            result = []

        if current.left:
            self.in_order(current.left, result)

        result.append(current.value)

        if current.right:
            self.in_order(current.right, result)

        return result

    def is_empty(self):
        return self.root is None

File: ruby_dsa/trees/bst/impl.rb
class TreeNode
  attr_accessor :value, :left, :right

  def initialize(value)
    @value = value
    @right = nil
    @left = nil
  end
end

class BST
  attr_reader :root

  def initialize
    @root = nil
  end

  def insert(value)
    new_node = TreeNode.new(value)

    return @root = new_node if empty?

    current = @root

    loop do
      if value < current.value
        current.left ? current = current.left : (break current.left = new_node)
      elsif value > current.value
        current.right ? current = current.right : (break current.right = new_node)
      else
        break
      end
    end
  end

  def search(value)
    return nil if empty?

    current = @root

    while current
      return true if value == current.value

      current = value < current.value ? current.left : current.right
    end

    false
  end

  def in_order(current = @root, result = [])
    return result if current.nil?

    in_order(current.left, result)
    result << current.value
    in_order(current.right, result)

    result
  end

  def pre_order(current = @root, result = [])
    return result if current.nil?

    result << current.value
    pre_order(current.left, result)
    pre_order(current.right, result)

    result
  end

  def post_order(current = @root, result = [])
    return result if current.nil?

    post_order(current.left, result)
    post_order(current.right, result)
    result << current.value

    result
  end

  def empty?
    root.nil?
  end
end
